### ISSUE TITLE
Improved address/interface binding

### ISSUE THREAD

**Issue by jeanpierrecartal** [esnet/iperf#1089]
# Context

* Version of iperf3:
iperf 3.9
* Hardware:
Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz / 16 GB RAM
* Operating system (and distribution, if any):
Linux 5.8.18-200.fc32.x86_64 (Fedora 32)

* Other relevant information (for example, non-default compilers,
  libraries, cross-compiling, etc.):

# Bug Report
-B option does not work properly with VLANs
* Expected Behavior
-B option should use the corresponding VLAN interface for outgoing packets
* Actual Behavior
The main interface is used with given IP as source.
* Steps to Reproduce
Create a VLAN with a different subnet, use -B option to specify the VLANs interface IP address as source.
e.g.:
```
$ ifconfig :
enp2s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.0.30  netmask 255.255.255.0  broadcast 192.168.0.255
...
enp2s0.5: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.5.5  netmask 255.255.255.0  broadcast 192.168.5.255
...
$ ./src/iperf3 -4 -c scaleway.testdebit.info -p 9210 -B 192.168.5.5
```
Using wireshark, no traffic is detected on interface enp2s0.5, traffic is seen on interface enp2s0 with 192.168.5.5 as source IP but no VLAN tagging is applied.

**Comment by bmah888** [comment]
I tried to reproduce this between two VirtualBox VMs on my laptop, using a VLAN created over a host-only virtual network. iperf3 seemed to behave correctly, with packets going between the two VMs and with the proper VLAN tags. (I seem to have tripped over a weird Virtualbox thing where I needed to have one or both of the network interfaces in promiscuous mode, such as with tcpdump, to receive VLAN-tagged packets.) Some more testing and/or thought will be needed to try to reproduce this problem.

In the example you gave, does 192.168.5.0/24 have a path to `scaleway.testdebit.info`?

**Comment by jeanpierrecartal** [comment]
The following command works properly, so the scaleway.testdebit.info server is reachable from this VLAN
```
$ curl -4 -o /dev/null --interface enp2s0.5 http://scaleway.testdebit.info/10G/10G.iso
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 28 9536M   28 2729M    0     0  74.5M      0  0:02:07  0:00:36  0:01:31 72.2M
```

**Comment by bmah888** [comment]
Thinking...sure I'll take a tcpdump file, thanks! I'm not sure yet how I'd use it but it'd be interesting to see the source and destination addresses. I don't need the whole transfer, maybe just give me a hundred packets or so.

For completeness, the server in my testing is FreeBSD 12.2 and the client is CentOS 8.

In your setup, if you do not explicitly bind to the tagged VLAN interface, would your test still work? If so, what interface would the test use, and what IP address would the client use for sending its packets?

I'm trying to visualize what's happening here and failing. It shouldn't be possible for the client machine to send IP packets with the source address on the VLAN interface, but out the untagged physical port. iperf3 doesn't even know that VLANs are involved...it's just using an IPv4 address bound to an interface that happens to be a VLAN interface.

**Comment by jeanpierrecartal** [comment]
tcpdump of curl command with --interface enp2s0.5 option:  [curl.dump.gz](https://github.com/esnet/iperf/files/5625583/curl.dump.gz)
tcpdump of iperf3 command with -B option: [vlan.dump.gz](https://github.com/esnet/iperf/files/5625587/vlan.dump.gz)
tcpdump of iperf3 command without the -B option: [default.dump.gz](https://github.com/esnet/iperf/files/5625605/default.dump.gz)

**Comment by jeanpierrecartal** [comment]
Looking at curl source code in lib/connect.c it is using the setsockopt(SO_BINDTODEVICE) to bind to the specified interface, removing this call makes curl behave like iperf.

**Comment by jeanpierrecartal** [comment]
According to my ncat bug report comments, this is an expected behaviour since the route to this host is through enp2s0 :
```
ip route get 62.210.156.7
62.210.156.7 via 192.168.0.254 dev enp2s0 src 192.168.0.30 uid 1001 
```
The iperf man page a little bit misleading for the -B option from my point of view.
Adding an option to behave like curl using the SO_BINDTODEVICE socket option would be a nice feature in my opinion.

**Comment by bmah888** [comment]
I'm going to go out on a limb and say this doesn't have anything to do with VLANs, but rather with address and interface bindings...changing the title of this issue to reflect this.

That socket option is Linux-specific (or at least it doesn't exist on every platform that we officially support)...I want to see how other OSs handle this issue, or if they even need to. Need to go study this a bit.

**Comment by bmah888** [comment]
I might need to dust off PR #817.  Also #450 is somewhat related to this.

**Comment by bmah888** [comment]
`issue-1089` branch holds some work in progress that was ported over from one of the commits making up #817.

**Comment by woky** [comment]
It seems that `--bind` combined with `--bind-dev` in version 3.10.1 still sends the traffic through my Ethernet USB dongle. I provide details below. `wlp2s0` is my laptop's wireless interface and `enp0s20f0u4u3` is gigabit Ethernet USB dongle. If I disconnect the dongle, the speed drops to cca 50 Mbps, probably running through wifi.

Version:
```
% iperf3 --version
iperf 3.10.1 (cJSON 1.7.13)
Linux bor 5.15.11-arch2-1 #1 SMP PREEMPT Wed, 22 Dec 2021 09:23:54 +0000 x86_64
Optional features available: CPU affinity setting, IPv6 flow label, TCP congestion algorithm setting, sendfile / zerocopy, socket pacing, authentication, bind to device, support IPv4 don't fragment
```

Interfaces:
```
% ip -4 a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
3: wlp2s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 192.168.34.50/24 brd 192.168.34.255 scope global dynamic noprefixroute wlp2s0
       valid_lft 38139sec preferred_lft 38139sec
7: enp0s20f0u4u3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    inet 192.168.34.114/24 brd 192.168.34.255 scope global dynamic noprefixroute enp0s20f0u4u3
       valid_lft 43075sec preferred_lft 43075sec
```

Test:
```
% sudo iperf3 --bind-dev wlp2s0 --bind 192.168.34.50 -c _gateway
Connecting to host _gateway, port 5201
[  5] local 192.168.34.50 port 59529 connected to 192.168.34.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   103 MBytes   863 Mbits/sec    0   1.67 MBytes       
[  5]   1.00-2.00   sec   112 MBytes   942 Mbits/sec    0   1.67 MBytes       
[  5]   2.00-3.00   sec   112 MBytes   943 Mbits/sec    0   1.75 MBytes       
^C[  5]   3.00-3.28   sec  31.2 MBytes   942 Mbits/sec    0   1.75 MBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-3.28   sec   359 MBytes   918 Mbits/sec    0             sender
[  5]   0.00-3.28   sec  0.00 Bytes  0.00 bits/sec                  receiver
iperf3: interrupt - the client has terminated
% 
```

**Comment by bmah888** [comment]
You might need the fix from PR #1153, which was merged relatively recently. It's not in 3.10.1, but if you can compile the code from the tip of the `master` branch it'll have that fix. If you need an actual release, 3.11 (planned for early 2022) will contain the fix also.
